backend_jobs/ingestion_pipeline/pipeline_lib/firestore_database.py

```diff
@@ -0,0 +1,147 @@
+  See the License for the specific language governing permissions and
+  limitations under the License.
+"""
+
+
+from datetime import datetime
+import random
+import apache_beam
+import geohash2
+from google.cloud import firestore
+from backend_jobs.pipeline_utils import firestore_database
+from backend_jobs.pipeline_utils import constants
+
+
+class StoreImageAttributeDoFn(apache_beam.DoFn):
+    """ Stores asynchronously the entity of ImageAttributes type from each image in the
+    projects database.
+    """
+    def setup(self):
+        self.database_firebase = firestore_database.initialize_db()
+
+    def process(self, element, provider, job_name):
+        """ Adds/Updates the projects database to contain documents with image attributes.
+        In addition it adds for each image a sub collection with information
+        on the pipelinerun and the provider.
+
+        Args:
+            element: ImageAttribute field
+            provider: the provider we are running now
+            job_name: the job name that we are running now
+        """
+        doc_ref = self.database_firebase \
+            .collection(constants.IMAGES_COLLECTION_NAME) \
+            .document(element.image_id)
+        doc = doc_ref.get()
+        sub_collection_ref = doc_ref.collection(constants.PIPELINE_RUNS_COLLECTION_NAME)
+        sub_collection_doc_ref = sub_collection_ref.document()
+        if doc.exists:
+            #doc found- image has been ingested already
+            _update_document(provider, doc, doc_ref, job_name)
+        else:
+            #doc not found- image has not been ingested already
+            _add_document(element, provider, job_name, doc_ref)
+        #Adding a doc to the sub collection (pipelinerun) in the image collection
+        _upload_sub_collection(element, provider, job_name, sub_collection_doc_ref)
+
+def _add_document(element, provider, job_name, doc_ref):
+    geo_point_coordinates = firestore.GeoPoint(
+        float(element.coordinates['latitude']),
+        float(element.coordinates['longitude']))
+    doc_ref.set({
+        constants.URL: element.url,
+        constants.INGESTED_PROVIDERS: [provider.provider_name],
+        constants.INGESTED_RUNS: [job_name],
+        constants.COORDINATES: geo_point_coordinates,
+        constants.DATE_INGESTED: datetime.now(),
+        constants.DATE_SHOT: element.date_shot,
+        constants.DATE_FIELDS:_get_date_fields(element.date_shot),
+        constants.HASHMAP: _get_geo_hashes_map(element.coordinates),
+        constants.IMAGE_ATTRIBUTES:{
+            constants.FORMAT: element.format,
+            constants.RESOLUTION:element.resolution},
+        constants.ATTRIBUTION: element.attribution,
+        constants.RANDOM: random.random(),
+        constants.VISIBILITY: provider.visibility,
+    })
+
+def _update_document(provider,doc, doc_ref,job_name):
+    ingested_runs = doc.to_dict()[constants.INGESTED_RUNS]
```

Comment:
what about the other fields? shall we update them? shall we check they're what we expect?

Reply:
I don't think we should update them ,
we can check them?
should I add a check function?